### PR TITLE
[storage] remove genesis in fast sync node

### PR DESCRIPTION
**Pull request by areshand**
### Description
The goal is to completely eliminate genesis from the DB in the fast-synced node.This change is the first step ONLY aims to solve the immediate issues of having genesis in the fast-synced DB. The next step of improvement should revise the interface between different components as mentioned below 

There are 2 steps to accomplish this change:
[IN this PR] Using a new struct for storing genesis DB and another DB for storing the fast-synced. The temporary DB and main DB is wrapped in a new struct that can be used both for bootstrapping and storing the actual data. This new struct providing all data interfaces and hide underly individual DB and selection logic from different components. 

[TODO] The next step is to restructure the interface between different components to rely on this new struct for status sync, mempool, event notification bootstrapping. Once the bootstrapping is done, we then create the actual underlying DB 
https://github.com/aptos-labs/aptos-core/pull/8849

By the way, Backup Handler currently depends on individual ledger, txn, and event stores. I've hidden these individual stores from the Backup Handler and exposed these underlying details through functions in the DB itself."

### Test Plan
Existing test

### CODE REVIEW

File: storage/aptosdb/src/backup/backup_data_accessor.rs
Comment (JoshLind) [inline storage/aptosdb/src/backup/backup_data_accessor.rs:2]: nit: This is a new file, so we don't need this in the license 😄

File: storage/aptosdb/src/backup/backup_data_accessor.rs
Comment (JoshLind) [inline storage/aptosdb/src/backup/backup_data_accessor.rs:21]: Silly question: why do we need this? If all calls go directly to the underlying wrapper DB, can we not just pass the wrapper to whatever objects require this?

I likely lack context here though 😄

File: storage/backup/backup-service/src/handlers/mod.rs
Comment (JoshLind) [inline storage/backup/backup-service/src/handlers/mod.rs:134]: Any context for these changes? They seem slightly orthogonal to me, e.g., different PR? 🤔

File: storage/backup/backup-service/src/lib.rs
Comment (JoshLind) [inline storage/backup/backup-service/src/lib.rs:18]: Silly question: I wonder if there's a way to avoid having to duplicate this, and instead pass in an interface, such that `start_backup_service` below can take in either the `FastSyncStorageWrapper` or the `AptosDB` implementation? Otherwise, it seems like we're doing a lot of duplication, no?

File: storage/storage-interface/src/lib.rs
Comment (JoshLind) [inline storage/storage-interface/src/lib.rs:656]: Silly question: why can we not place this in the `FastSyncStorageWrapper` implementation itself? Putting it here feels off to me, i.e., we're leaking implementation details. Theoretically, `reader` and `writer` both take in interfaces, which should be `FastSyncStorageWrapper` underneath, right?

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (JoshLind) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:2]: nit: This is a new file, so we don't need this in the license 😄

File: storage/storage-interface/src/lib.rs
Comment (JoshLind) [reply storage/storage-interface/src/lib.rs:656]: Actually, I see that it's already in there. Why do we need it here? 🤔

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (JoshLind) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:59]: Interesting... some thoughts:

1. Nit: I'd likely call these something more obvious, e.g.: `temporary_db_with_genesis` and `real_db_for_fast_sync`?
2. I probably lack context, but do these types need to be `AptosDB`? Why can they not be `DBReaderWriter`? Is this because of the backup system?
3. If we only use this wrapper in the context of fast sync, you can avoid the `Option<AptosDB>` (for the second DB) and clean up the unwraps below?
4. I don't really know why we need `FastSyncStatus`? Could we not just have a `fast_sync_complete: AtomicBool`? This would avoid the need for grabbing a RWLock for all reads and writes, and simplify the switches (i.e., the `FastSyncStatus::UNKNOWN` and `FastSyncStatus::STARTED` doesn't mean much, right? All we need to know is when it finishes, so we can direct all calls to `real_db_for_fast_sync`?) Does that make sense?

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (JoshLind) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:435]: Eventually, I could imagine us defining a macro for this (if we wanted to clean things up -- especially since all the switch statements are identical 😄 )

**Review by JoshLind — COMMENTED** [review]
Thanks @areshand! Looks like a solid first pass. 😄  Most of my feedback is about code structure, abstractions and simplifications. Let me know if anything doesn't make sense.

File: state-sync/state-sync-v2/state-sync-driver/src/storage_synchronizer.rs
Comment (areshand) [reply state-sync/state-sync-v2/state-sync-driver/src/storage_synchronizer.rs:113]: The reason is the status (fast_sync_status) has to be stored in a struct (DbReaderWriter). However, get_state_snapshot_receiver and get_state_snapshot_receiver don't have access to db_rw, which doesn't have access to underlying status flag.

File: storage/backup/backup-service/src/lib.rs
Comment (areshand) [reply storage/backup/backup-service/src/lib.rs:18]: will remove `start_backup_service` which is itself a lot of work and want to only delete it once this new function is accepted.

File: storage/storage-interface/src/lib.rs
Comment (areshand) [reply storage/storage-interface/src/lib.rs:656]: DbReaderWriter is the `interface` struct, which state-sync and other components use to access storage. State-sync is not supposed to directly access underlying FastSyncStorageWrapper's fields.

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (areshand) [reply storage/aptosdb/src/fast_sync_aptos_db.rs:59]: 1. that was my previous implementation. But there is a corner case. For example, the node can be configured to do fast sync. When all its peers only have genesis data, it would actually do a continuous update of txns one by one. Storage in this case should put the genesis and following updates in the same DB.  
2. Yes, this is related to the backup system since it needs to access the underlying txn_store, event_store, etc of the DB. 
3. I don't know if that makes the logic more complicated. since you need to remember AptosDB is used in this circumstance and FastSyncDBWrapper is used in other circumstances. For now, I just make FastSyncDBWrapper used in all circumstances and people can check one place to find the code behavior.
4. as mentioned in 1. we need to know if fast sync indeed starts to determine if we should use a separate DB. the `Unknown` is used for indicating non-fast-sync cases.

File: storage/backup/backup-service/src/handlers/mod.rs
Comment (areshand) [reply storage/backup/backup-service/src/handlers/mod.rs:134]: related. backup service needs to know the sub dbs (event, txn, state store, etc).  When we switch DB based on Fast Sync status, the backup handler service will also need to switch the DB

File: storage/aptosdb/src/backup/backup_data_accessor.rs
Comment (areshand) [reply storage/aptosdb/src/backup/backup_data_accessor.rs:21]: yes, we can directly use the DBWrapper for Backup Service to avoid introducing backup_data_acessor. I was just following the convention of BackupHandler

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (msmouse) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:132]: Too much duplication. We can at least abstract this piece out as `fn get_aptosdb_db_ref() -> &AptosDB`.  I would go as far as doing a 
```
trait GetAptosDbRef {
    fn get_aptos_db_ref() -> &AptosDB;
}
```
And implement that for both the actual AptosDB and the wrapper, and then
```
impl<T: GetAptosDbRef> DbWriter for T {
...
}
```

Comment (msmouse) [reply storage/aptosdb/src/fast_sync_aptos_db.rs:132]: Or darker (but elegant) magic like https://github.com/aptos-labs/aptos-core/blob/547ad67a0126f9447ef9aaf19863a56203a3e17d/aptos-move/aptos-gas-profiling/src/profiler.rs#L51

File: aptos-node/src/storage.rs
Comment (areshand) [reply aptos-node/src/storage.rs:26]: I also thought of this but didn't find a good way.  eventually, I think it is better to have all the logic inside FastSyncStorageWrapper instead of exposing the complexity of selecting different DB inside aptos-node. Eg: the fast sync storage wrapper will decide not to create the secondary DB if the config is not turned on. But happy to know if there is a better way to do this.

Comment (areshand) [reply aptos-node/src/storage.rs:37]: I also thought of this but didn't find a good way.  eventually, I think it is better to have all the logic inside FastSyncStorageWrapper instead of exposing the complexity of selecting different DB inside aptos-node. Eg: the fast sync storage wrapper will decide not to create the secondary DB if the config is not turned on. But happy to know if there is a better way to do this.

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (JoshLind) [reply storage/aptosdb/src/fast_sync_aptos_db.rs:54]: 1. I'd need to look at the state sync code again, but if that's the case, I think we should change state sync such that it will always fast sync (i.e., if it starts up in fast sync mode, and we haven't had a single epoch change, state sync will just loop until the first epoch change occurs). Otherwise, this feels a bit janky to me.
3 + 4. Not really following you -- will sync offline 😄

File: storage/storage-interface/src/lib.rs
Comment (JoshLind) [reply storage/storage-interface/src/lib.rs:656]: Still not following -- will sync offline 😄

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (areshand) [reply storage/aptosdb/src/fast_sync_aptos_db.rs:54]: updated based on offline discussion. The test will be correct once the state sync change is landed.

File: aptos-node/src/storage.rs
Comment (areshand) [reply aptos-node/src/storage.rs]: updated based on offline discussion

File: storage/storage-interface/src/lib.rs
Comment (JoshLind) [inline storage/storage-interface/src/lib.rs:642]: nit: can this live in the fast sync storage wrapper file?

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (JoshLind) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:389]: nit: any chance we could still have empty lines between the functions to make it more readable? 😄

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (JoshLind) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:53]: nit: can we update this comment? Also, do you think it makes sense to rename these now? e.g., `temporary_db_with_genesis` and `real_db_for_fast_sync`?

**Review by JoshLind — COMMENTED** [review]
Thanks @areshand! This looks much cleaner to me 😄

I'll send out a PR to update state sync and then approve this PR! We should still wait for the storage experts to review this to make sure they're happy. But, otherwise, this LGTM.

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (areshand) [reply storage/aptosdb/src/fast_sync_aptos_db.rs:53]: updated

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (msmouse) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:86]: "configured"

File: aptos-node/src/storage.rs
Comment (areshand) [reply aptos-node/src/storage.rs:42]: we already have Aptoshandle taking a backup runtime. why do we want to do this?

Comment (msmouse) [reply aptos-node/src/storage.rs:42]: To remove the need for the backup service to be able to handle both the db and the wrapper. Our end goal is to spin up all the components that are not necessary during bootstrapping only after the  bootstrapping is finished, right? the backup service is such a component.
(The runtime on the handle is optional, no need to change anything on that front.)

Comment (areshand) [reply aptos-node/src/storage.rs:42]: The FastSyncStorageWrapper is only created for fast sync. For non-fast sync, I don't have a wrapper to hold the runtime. Similarly, backup service just need to know either DB or FastSyncWrapper. However, FastSyncWrapper is currently only for Fast sync not for other non-fast sync cases. Not sure how to do this refactor.

Comment (msmouse) [reply aptos-node/src/storage.rs]: > For non-fast sync, I don't have a wrapper to hold the runtime
the aptos handle holds it in that case -- that's why I said it's convenient that it's an `Option<Runtime>`

it's a bit awkward to have to hold the runtime in different places, but you don't need to copy paste a few functions as you do today. And at high level it's the right thing to do: spin up the backup service only after bootstrapping is done.

**Review by msmouse — APPROVED** [review]

File: aptos-node/src/storage.rs
Comment (msmouse) [inline aptos-node/src/storage.rs:52]: This will write to the real DB instead of the temp DB? Will that even work?

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (JoshLind) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:44]: Remove?

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (JoshLind) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:58]: nit: you should be able to call:

```if config.state_sync.state_sync_driver.bootstrapping_mode.is_fast_sync()```

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (JoshLind) [inline storage/aptosdb/src/fast_sync_aptos_db.rs:19]: nit: why not `use aptos_infallible::RwLock;`? You can get rid of the `expects` below.

**Review by JoshLind — APPROVED** [review]
Looks clean 😄 Thanks, all!

File: storage/aptosdb/src/fast_sync_aptos_db.rs
Comment (msmouse) [reply storage/aptosdb/src/fast_sync_aptos_db.rs:19]: :trollface:  missed this one

**Pull request merged**
